subsys/bluetooth/host/l2cap.c

```diff
@@ -1355,6 +1360,17 @@ static void l2cap_chan_update_credits(struct bt_l2cap_le_chan *chan,
 	BT_DBG("chan %p credits %u", chan, k_sem_count_get(&chan->rx.credits));
 }
 
+u16_t bt_l2cap_chan_set_credits(struct bt_l2cap_chan *chan, u16_t credits)
```

Comment:
It is confusing to me that the param is called credits when what the function does is to set init_credits, i.e. you are not giving credits in a relative way (like LE Flow Control Credit packet). As hinted above, I'm unsure of how to use this function in practice.

Reply:
It is used to restore the credits, though perhaps it clearer if we do say init_credits instead, we cannot take back credits so that can only affect the next time we update the credits so you call it whenever you want to change the init_credits while connected e.g. if you run out of buffer set it to 0 which will likely make the remote stop to transmit, then as the buffers return to the pool you may restore the credits.

Reply:
The credits are counted in L2CAP MPS (PDU), but the alloc_buf callback is supposed to allocate an SDU (L2CAP MTU). So, how will I now how many credits to give after I set init_credits=0?
Further, can I be sure not to get a recv() callback after setting init_credits=0, i.e. could there be packets in the pipeline?

The way it is done in NimBLE seems like a good design, where the credit is hidden from the application. The application just calls 
````
ble_l2cap_recv_ready(*chan, sdu_rx) 
````
when it is able to receive another SDU. Until then, the stack buffers as much as its pool allows it.

It may still be a good idea to keep the set_init_credits in order to dynamically re-prioritize the buffer usage to different channels.

 https://github.com/apache/mynewt-nimble/blob/1f7227ed860c8783ec874effde4c0bd620977187/nimble/host/src/ble_l2cap_coc.c#L449
https://www.mail-archive.com/dev@mynewt.incubator.apache.org/msg02418.html